File: scripts/lib/output.js
const { OUTPUT_FORMATS, ERROR_CODES } = require('./constants');
const { AppError } = require('./errors');

function toDisplayValue(value) {
  if (value === null || value === undefined) {
    return '';
  }
  if (typeof value === 'object') {
    return JSON.stringify(value);
  }
  return String(value);
}

function toTable(rows) {
  if (!Array.isArray(rows) || rows.length === 0) {
    return 'No rows.';
  }

  const normalizedRows = rows.map((row) => {
    if (row && typeof row === 'object' && !Array.isArray(row)) {
      return row;
    }
    return { value: row };
  });

  const columns = Array.from(
    normalizedRows.reduce((set, row) => {
      for (const key of Object.keys(row)) {
        set.add(key);
      }
      return set;
    }, new Set())
  );

  const widths = columns.map((column) =>
    Math.max(column.length, ...normalizedRows.map((row) => toDisplayValue(row[column]).length))
  );

  const header = columns
    .map((column, index) => column.padEnd(widths[index], ' '))
    .join(' | ');
  const separator = widths.map((width) => '-'.repeat(width)).join('-|-');
  const body = normalizedRows
    .map((row) =>
      columns
        .map((column, index) => toDisplayValue(row[column]).padEnd(widths[index], ' '))
        .join(' | ')
    )
    .join('\n');

  return `${header}\n${separator}\n${body}`;
}

function printResult(result, outputFormat) {
  if (!OUTPUT_FORMATS.includes(outputFormat)) {
    throw new AppError(
      ERROR_CODES.VALIDATION,
      `Invalid output format "${outputFormat}". Allowed values: ${OUTPUT_FORMATS.join(', ')}.`
    );
  }

  if (outputFormat === 'table') {
    if (result && Array.isArray(result.tableRows)) {
      console.log(toTable(result.tableRows));
      return;
    }
    if (Array.isArray(result)) {
      console.log(toTable(result));
      return;
    }
  }

  const payload = result && Object.prototype.hasOwnProperty.call(result, 'data') ? result.data : result;
  console.log(JSON.stringify(payload, null, 2));
}

function printError(error) {
  const payload = {
    ok: false,
    error: {
      code: error.code,
      message: error.message
    }
  };
  if (error.status !== undefined) {
    payload.error.status = error.status;
  }
  if (error.details !== undefined) {
    payload.error.details = error.details;
  }
  console.error(JSON.stringify(payload, null, 2));
}

module.exports = {
  printResult,
  printError,
  toTable
};

File: scripts/python/lib/output.py
import json
import sys
from .constants import OUTPUT_FORMATS, ERROR_CODES
from .errors import AppError

def to_display_value(value):
    if value is None:
        return ''
    if isinstance(value, (dict, list)):
        return json.dumps(value)
    return str(value)

def to_table(rows):
    if not isinstance(rows, list) or len(rows) == 0:
        return 'No rows.'

    normalized_rows = []
    for row in rows:
        if isinstance(row, dict):
            normalized_rows.append(row)
        else:
            normalized_rows.append({'value': row})

    columns = []
    seen_columns = set()
    for row in normalized_rows:
        for key in row.keys():
            if key not in seen_columns:
                columns.append(key)
                seen_columns.add(key)

    widths = []
    for column in columns:
        max_w = len(column)
        for row in normalized_rows:
            val_w = len(to_display_value(row.get(column)))
            if val_w > max_w:
                max_w = val_w
        widths.append(max_w)

    header_parts = []
    for i, column in enumerate(columns):
        header_parts.append(column.ljust(widths[i]))
    header = ' | '.join(header_parts)

    separator_parts = []
    for width in widths:
        separator_parts.append('-' * width)
    separator = '-|-'.join(separator_parts)

    body_lines = []
    for row in normalized_rows:
        row_parts = []
        for i, column in enumerate(columns):
            row_parts.append(to_display_value(row.get(column)).ljust(widths[i]))
        body_lines.append(' | '.join(row_parts))
    body = '\n'.join(body_lines)

    return f"{header}\n{separator}\n{body}"

def print_result(result, output_format):
    if output_format not in OUTPUT_FORMATS:
        raise AppError(
            ERROR_CODES.VALIDATION,
            f"Invalid output format \"{output_format}\". Allowed values: {', '.join(OUTPUT_FORMATS)}."
        )

    if output_format == 'table':
        if isinstance(result, dict) and isinstance(result.get('tableRows'), list):
            print(to_table(result['tableRows']))
            return
        if isinstance(result, list):
            print(to_table(result))
            return

    payload = result.get('data') if isinstance(result, dict) and 'data' in result else result
    print(json.dumps(payload, indent=2))

def print_error(error):
    payload = {
        'ok': False,
        'error': {
            'code': getattr(error, 'code', ERROR_CODES.UNKNOWN),
            'message': str(error)
        }
    }
    if hasattr(error, 'status') and error.status is not None:
        payload['error']['status'] = error.status
    if hasattr(error, 'details') and error.details is not None:
        payload['error']['details'] = error.details
    
    print(json.dumps(payload, indent=2), file=sys.stderr)
